go/launcher/proxy/driverhub/debugger/debugger.go

```diff
@@ -0,0 +1,274 @@
+func (d *Debugger) Request(r *http.Request) {
+  d.mu.RLock()
+  step := d.step
+
+  if !step {
+    for _, bp := range d.breakpoints {
+      if bp.matches(r) {
+        step = true
+        break
+      }
+    }
+  }
+
+  d.mu.RUnlock()
+
+  resp := &response{
+    Request: request{
+      Method: r.Method,
+      Path:   r.URL.Path,
+    },
+  }
+
+  if step {
+    resp.Status = "waiting"
+  } else {
+    resp.Status = "running"
+  }
+
+  bytes, err := json.Marshal(resp)
+  if err != nil {
+    log.Print(err)
+    return
+  }
+
+  if _, err := d.conn.Write(bytes); err != nil {
+    log.Print(err)
+  }
+
+  if _, err := d.conn.Write([]byte("\n")); err != nil {
+    log.Print(err)
+  }
+
+  if !step {
+    return
+  }
+
+  // TODO(DrMarcII): fix race condition...
+  waiting := make(chan interface{})
+  d.mu.Lock()
+  d.waiting = waiting
+  d.mu.Unlock()
+
+  <-waiting
+  d.mu.Lock()
+  d.waiting = nil
+  d.mu.Unlock()
+}
+
+func (d *Debugger) waitForConnection(port int) {
+  l, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+  if err != nil {
+    d.mu.Lock()
+    defer d.mu.Unlock()
+    d.failed = errors.NewPermanent(d.Name(), err)
+    return
+  }
+
+  fmt.Printf("Waiting for connect on port %d\n", port)
+
+  conn, err := l.Accept()
+  d.mu.Lock()
+  defer d.mu.Unlock()
+  if err != nil {
+    d.failed = errors.NewPermanent(d.Name(), err)
+    return
+  }
+  d.conn = conn
+
+  go d.readLoop()
+}
+
+func (d *Debugger) readLoop() {
+  decoder := json.NewDecoder(d.conn)
+
+  for {
+    cmd := &command{}
+    if err := decoder.Decode(cmd); err != nil {
+      // do something here...
+      log.Print(err)
+      continue
+    }
+
+    d.processCommand(cmd)
+  }
+}
+
+func (d *Debugger) processCommand(cmd *command) {
+  d.mu.Lock()
+  defer d.mu.Unlock()
+
+  fmt.Printf("Cmd: %+v\n", cmd)
+
+  response := &response{ID: cmd.ID, Status: "error"}
+
+  switch cmd.Command {
+  case "continue":
+    d.healthy = true
+    d.step = false
+    if d.waiting != nil {
+      close(d.waiting)
+    }
+    response.Status = "running"
+
+  case "step":
+    d.healthy = true
+    d.step = true
+    if d.waiting != nil {
+      close(d.waiting)
+    }
+    response.Status = "running"
+
+  case "stop":
+    os.Exit(-1)
+  case "set breakpoint":
+    if cmd.Breakpoint == nil {
+      break
+    }
+    bp := cmd.Breakpoint
+    if bp.Path != "" {
+      r, err := regexp.Compile(bp.Path)
+      if err != nil {
+        break
+      }
+      bp.pathRegex = r
+    }
+    d.breakpoints[bp.ID] = bp
+    response.Status = "waiting"
+
+  case "delete breakpoint":
+    if cmd.Breakpoint == nil {
+      break
+    }
+    delete(d.breakpoints, cmd.Breakpoint.ID)
+    response.Status = "waiting"
+  }
+
+  bytes, err := json.Marshal(response)
+  if err != nil {
+    log.Print(err)
+    return
+  }
+
+  if _, err := d.conn.Write(bytes); err != nil {
+    log.Print(err)
+  }
+}
+
+func (bp *breakpoint) matches(r *http.Request) bool {
```

Comment:
It would be nice to have some tests for the breakpoint matching logic. E.g., an empty path/method matches every path/method (and therefore, the zero-value breakpoint matches every command).